doc/specs/stdlib_hash_maps.md

```diff
@@ -0,0 +1,3088 @@
+the ratio of the number of hash map probes to the number of subroutine 
+calls.
+The maps make extensive use of pointers internally, but a private
+finalization subroutine avoids memory leaks.
+The maps can take entry keys of type `key_type`.
+Both maps allow the addition and lookup of entries, and the inclusion
+of data in addition to the entry key.
+The `chaining_hash_map_type` also allows the selective removal of
+entries.
+
+## The `stdlib_32_bit_key_data_wrapper` module
+
+The `stdlib_32_bit_key_data_wrapper` module provides data types to
+represent keys and associated data stored in a module, but is also, a
+wrapper for the `stdlib_32_bit_hash_functions` module. It allows
+direct access to the `stdlib_32_bit_hash_functions` procedures:
+`fibonacci_hash`, `fnv_1_hasher`, `fnv_1a_hasher`; and provides
+wrapper functions, `seeded_nmhash32_hasher`,
+`seeded_nmhash32x_hasher`, and `seeded_water_hasher` to the hashing
+functions: `nmhash32`, `nmhash32x`, and `water_hash`, respectively. It
+defines an interface, `hasher_fun`, compatible with the hash functions
+that take a `non-scalar key`. It defines one integer constant used
+as a kind value,`int_hash`. It also defines two types, `key_type` and
+`other_type`, and associated procedures, for storing and manipulating
+keys and their associated data.
+
+### The `stdlib_32_bit_key_data_wrapper` constant, `INT_HASH`
+
+The constant `INT_HASH` is used to define the integer kind value for
+the returned hash codes and variables used to access them. It
+currently has the value, `INT32`.
+
+### The `stdlib_32_bit_key_data_wrapper` module derived types
+
+The `stdlib_32_bit_key_data_wrapper` module defines two derived types:
+`key_type`, and `other_type`. The `key_type` is intended to be used
+for the search keys of hash tables.  The `other_type` is intended to
+store additional data associated with a key. Both types are
+opaque. Their current representations are as follows
+
+```fortran
+    type :: key_type
+        private
+        integer(int8), allocatable :: value(:)
+    end type key_type
+
+    type :: other_type
+        private
+        integer(int8), allocatable :: value(:)
+    end type other_type
+```
+
+The  module also defines seven procedures for those types: `copy_key`,
+`copy_other`, `free_key`, `free_other`, `get`, `key_test`, and `set`
+for use by the hash maps to manipulate or inquire of components of
+those types.
+
+### Table of `stdlib_32_bit_key_data_wrapper` procedures
+
+The  `stdlib_32_bit_key_data_wrapper` module provides procedures in
+several categories: procedures to manipulate data of the `key_type`;
+procedures to manipulate data of the `other_type`, and 32 bit hash
+functions for keys. The procedures in each category are listed below.
+
+Procedures to manipulate `key_type` data:
+
+* `copy_key( key_in, key_out )` - Copies the contents of the key,
+  key_in, to the key, key_out.
+
+* `get( key, value )` - extracts the content of key into value.
```

Comment:
Not immediately clear to me how this differs from `copy_key`

Reply:
I see the details are explained below -- some minor edits here should make the difference clear.

Reply:
`copy_key` copies the internals of one key to the internals of another key. `get` copies the contents of a key or other data into an allocatable default character string or `int8` array. I have improved the documentation of this in `stdlib_hash_maps.md`.

Reply:
I added a few more details and pushed the result.